### DISCUSSION TITLE
`useHydrated` and server streaming (defer)

### DISCUSSION THREAD

**Discussion by maranomynet** [sergiodxa/remix-utils#205] [Q&A]
I'm curious. How does `useHydrated` play with streamed server responses, where parts of the page are deferred?

Are the suspended parts not "hydrated" in the same way as the initial batch of HTML?
And if so, will React issue hydration warnings if useHydrated returns its internally stored `true` value?

**Comment by sergiodxa** [comment]
useHydrated works by changing a Boolean inside an Effect.

If the component using useHydrated is not Suspended it will hydrate even before the rest of the app is unsuspended.

If all components using useHydrated are render inside a Suspense boundary and are suspended then once they unsuspend it will first return false and later change to true.

If you have a mix of both, then the unsuspended component will cause useHydrated to change to true and the suspended component once ready will get true immediately.

**Reply by maranomynet** [reply]
Thanks. One question though: In the third/last case, will there be no hydration errors/warnings when the suspended server content arrives and hydrates with the useHydrated boolean already being true?
Does suspended content hydrate differently from unsuspended server rendered HTML?

**Reply by sergiodxa** [reply, marked as answer]
It will be the same html in both cases, actually it will render once client side

**Reply by maranomynet** [reply]
Thank you for this clarification.

**Reply by maranomynet** [reply]
> Thanks. One question though: In the third/last case, will there be no hydration errors/warnings when the suspended server content arrives and hydrates with the useHydrated boolean already being true?

@sergiodxa I'm actually seeing hydration warnings in react-router (v7.18) when using `useHydrated()` inside `<Await/>`.

<img width="600" height="285" alt="image" src="https://github.com/user-attachments/assets/1bc989f8-7c5e-4b40-8b43-ebb101dfe81a" />

My `meanwhile={}` component is like this:

```ts
import { useHydrated } from 'remix-utils/use-hydrated';

const MeanWhile = () => {
  const hyrdrating = useHydrated();
  return hyrdrating ? <p>Loading...</p> : <p>Server rendered fallback...</p>;
};
```

…and the same hydration warning appears, of course, if I use a similar component in the `<Await/>` render function.

Any idea how to resolve this nicely?

**Reply by maranomynet** [reply]
Converted to an issue https://github.com/sergiodxa/remix-utils/issues/613
